libopenage/input/input_manager.cpp

```diff
@@ -12,57 +12,139 @@ namespace input {
+	      { sdl_key(SDLK_e, KMOD_LCTRL), actions::ACTION_MAP.at("BUILDING_7") },
+	      { sdl_key(SDLK_r, KMOD_LCTRL), actions::ACTION_MAP.at("BUILDING_8") },
+	      { sdl_key(SDLK_z), actions::ACTION_MAP.at("BUILDING_TOWN_CENTER") },
+	      { sdl_key(SDLK_1), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_1") },
+	      { sdl_key(SDLK_2), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_2") },
+	      { sdl_key(SDLK_3), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_3") },
+	      { sdl_key(SDLK_4), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_4") },
+	      { sdl_key(SDLK_5), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_5") },
+	      { sdl_key(SDLK_6), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_6") },
+	      { sdl_key(SDLK_7), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_7") },
+	      { sdl_key(SDLK_8), actions::ACTION_MAP.at("SWITCH_TO_PLAYER_8") },
+	      { sdl_key(SDLK_UP), actions::ACTION_MAP.at("UP_ARROW") },
+	      { sdl_key(SDLK_DOWN), actions::ACTION_MAP.at("DOWN_ARROW") },
+	      { sdl_key(SDLK_LEFT), actions::ACTION_MAP.at("LEFT_ARROW") },
+	      { sdl_key(SDLK_RIGHT), actions::ACTION_MAP.at("RIGHT_ARROW") },
+	      { sdl_mouse(1), actions::ACTION_MAP.at("PAINT_TERRAIN") },
+	      { sdl_mouse(1), actions::ACTION_MAP.at("SELECT") },
+	      { sdl_mouse(3), actions::ACTION_MAP.at("DESELECT") },
+	      { sdl_wheel(1), actions::ACTION_MAP.at("FORWARD") },
+	      { sdl_wheel(-1), actions::ACTION_MAP.at("BACK") }},
 	relative_mode{false} {
 }
 
+std::string InputManager::get_key(const std::string action_str) {
+	for (auto it=keys.begin(); it!=keys.end(); ++it) {
+		if (it->second.info == action_str) {
+			std::string key_str = std::string(SDL_GetKeyName(it->first.cc.code));
+
+			auto end = it->first.mod.end();
+			if (it->first.mod.find(modifier::ALT) != end) {
+				key_str = "Alt "+key_str;
+			}
+			if (it->first.mod.find(modifier::SHIFT) != end) {
+				key_str = "Shift "+key_str;
+			}
+			if (it->first.mod.find(modifier::CTRL) != end) {
+				key_str = "Ctrl "+key_str;
+			}
+
+			return key_str;
+		}
+	}
+	return "";
+}
+
+bool InputManager::set_key(const char* key_char, const std::string action_str) {
+	try {
+		Event ev = textToEvent(key_char);
+
+		for (auto it=keys.begin(); it!=keys.end(); ++it) {
+			if (it->second.info == action_str) {
+				keys.erase(it);
+				break;
+			}
+		}
+
+		auto it = actions::ACTION_MAP.find(action_str);
+		if (it == actions::ACTION_MAP.end())
+			return false;
+		keys.emplace(ev, it->second);
+		return true;
+	}
+	catch (int error) {
+		return false;
+	}
+}
+
+Event InputManager::textToEvent(const char* key_char)
+{
+	std::string key_str = std::string(key_char);
+	std::size_t last = key_str.rfind(" ");
+
+	if (last == std::string::npos) {
+		SDL_Keycode key_code = SDL_GetKeyFromName(key_char);
+		if (key_code == SDLK_UNKNOWN) {
+			throw 0;
+		}
+		return sdl_key(key_code);
+	}
+
+	SDL_Keycode key_code = SDL_GetKeyFromName(key_str.substr(last+1).c_str());
+	if (key_code == SDLK_UNKNOWN) {
+			throw 0;
+	}
+	int key_mod = 0;
```

Comment:
How can this happen? Is this the splitup at `" "` to allow more than one key? If yes, it can be implemented shorter I think, without the copy-paste of `GetKeyFromName` and the check.